spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseExpireSnapshotsSparkAction.java

```diff
@@ -212,11 +214,21 @@ private String jobDesc() {
   }
 
   private ExpireSnapshots.Result doExecute() {
-    boolean streamResults = PropertyUtil.propertyAsBoolean(options(), STREAM_RESULTS, false);
-    if (streamResults) {
-      return deleteFiles(expire().toLocalIterator());
+    boolean parallelDelete = PropertyUtil.propertyAsBoolean(options(), PARALLEL_DELETE, false);
+
+    if (parallelDelete) {
+      Table serializableTable = SerializableTable.copyOf(table);
+
+      return expire().javaRDD().mapPartitions((Iterator<Row> iter) ->
+        Arrays.asList(
+            ActionUtils.deleteFiles(iter, (file) -> serializableTable.io().deleteFile(file), null)
+        ).iterator()
+      ).reduce((r1, r2) -> new BaseExpireSnapshotsActionResult(
+          r1.deletedDataFilesCount() + r2.deletedDataFilesCount(),
+          r1.deletedManifestsCount() + r2.deletedManifestsCount(),
+          r1.deletedManifestListsCount() + r2.deletedManifestListsCount()));
```

Comment:
Using Spark to delete files can easily result in too high of a request rate, which causes S3 to throttle and send 503, "Slow down", responses. When that happens, files are probably leaked as orphans. That's why we have a way to parallelize locally, but not to use Spark.

This action also exposes the `expire()` method that returns a `Dataset` you can use if you want to run the expiration like this in your environment.

Reply:
@rdblue Thanks for your comment. In our last few runs, we didn't see too many S3 503 errors, maybe the requests were distributed into different prefixes(with partition dt/hh). I think by using the local parallelize mechanism, it might require more resource on driver side, while all the executors seem to be all in idle. E.g. I need to request more CPUs for the driver to be able to view the Spark UI(50 concurrent delete), which is not cost efficient considering it also needs longer time to run.

For the request with 503, we probably can collect them back into the driver for anther retry?